src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2785,17 +2790,34 @@ internal bool ShouldSuppressAnalysisWarningsForRequiresUnreferencedCode ()
 
 			IMemberDefinition suppressionContextMember = currentOrigin.SuppressionContextMember;
 			if (suppressionContextMember != null &&
-				Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (suppressionContextMember))
+				(Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (suppressionContextMember) ||
+				(suppressionContextMember.DeclaringType != null &&
+				Annotations.HasRequiresUnreferencedCodeOnTypeHierarchy (suppressionContextMember.DeclaringType))))
 				return true;
 
 			IMemberDefinition originMember = currentOrigin.MemberDefinition;
 			if (suppressionContextMember != originMember && originMember != null &&
-				Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (originMember))
+				(Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (originMember) ||
+				(originMember.DeclaringType != null &&
+				Annotations.HasRequiresUnreferencedCodeOnTypeHierarchy (originMember.DeclaringType))))
 				return true;
 
 			return false;
 		}
 
+		internal void CheckAndReportRequiresUnreferencedCode (TypeDefinition type)
+		{
+			var currentOrigin = _scopeStack.CurrentScope.Origin;
+
+			// If the caller of a type is already marked with `RequiresUnreferencedCodeAttribute` a new warning should not
+			// be produced for the callee.
+			if (ShouldSuppressAnalysisWarningsForRequiresUnreferencedCode ())
```

Comment:
This will probably also not handle suppressions correctly since the current scope will not be right.

Reply:
I have some scenarios with suppressions that are working fine, although is not extensive. Do you have a particular scenario that you would like me to test?